third/lprng/lprng/man/lpc.1
.ds VE LPRng-3.5.3
.TH LPC 1 \*(VE "LPRng"
.ig
lpc.1,v 3.36 1998/03/29 18:37:49 papowell Exp
Mon Jul 17 20:14:17 PDT 1995 Patrick Powell
..
.SH NAME
lpc \- line printer control program
.SH SYNOPSIS
.B lpc
[
.B \-A
]
[
.BI \-D debugopts
]
[
.BI \-P printer
]
[
.B \-V
]
[ command [ argument ... ] ]
.SH DESCRIPTION
.I Lpc
is used by the system administrator to control the
operation of the line printer system.  
For each line printer configured in 
.I /etc/printcap 
or the printcaps specified in the configuration file 
.I lpd.conf
(see lpd.conf(5)),
.I lpc
may be used to:
.IP \(bu 3
disable or enable a single or all printer,
.IP \(bu 3
disable or enable a single printer's or all printer's spooling queue,
.IP \(bu 3
rearrange the order of jobs in a spooling queue (moving jobs to the
top or to the end of a queue or changing their priority in the queue),
.IP \(bu 3
find the status of printers, and their associated
spooling queues and printer dameons,
.IP \(bu 3
start and stop printer servers for a queue with multiple print servers,
.IP \(bu 3
hold and release a specific job in a printer queue,
.IP \(bu 3
redirect printing to another printer,
.IP \(bu 3
restart a printer job after having solved printer problems and let it
print from the beginning, 
.IP \(bu 3
reprint a job
.PP
Without any arguments,
.I lpc
will prompt for commands from the standard input.
If arguments are supplied,
.IR lpc
interprets the first argument as a command and the remaining
arguments as parameters to the command.
Permission to use spool queue control commands is determined by the
printer permissions file
(See FILES).
.SH OPERATION
.PP
The operation of each spool queue is controlled by a
.I "spoolcontrol"
file which has a set of keyword options and values.
(Other print spoolers have used permissions bits of directories and files
for similar purposes).
In addition,
each job in the spool queue can have a
.I "holdfile"
which contains detailed information on how the spooler is to treat the
particular job.
The server will update these files with status and other information as the
job is processed.
The
.I lpc
command operates by sending requests to the
.I lpd
server process to update the information in these files and to signal
server processes that the information has been updated.
.SH OPTIONS
.IP "\fB\-A\fR" 5
Use authentication when sending command to server.
.IP "\fB\-P\fIprinter\fR" 5
Printer spool queue to operate on.
if no name is given,
the -P option,
the printer selected by the value of the PRINTER enviroment variable,
or the first entry in the printcap file will be used.
When all printers are selected,
the LPC command will first attempt to find the
printcap entry for printer ``all'';
the
.I all
printcap field value will be a list of printers to control. 
If there is no printcap entry,
then the request is directly forwarded to the default LPD host,
which will do a lookup for all of the available printers.
By specifying ``all@host'',
the user can control all printers on a given host.
.IP "\fB\-V\fR" 5
Print program version information
.PP
.IP "\fB\-D\fIdebugopts\fP" 5
Debugging is controlled using the
.B \-D
option. This accepts a comma-separated list of debugging
settings. These settings take one of two forms: 
.B facility=value , 
or
.B value 
to set an overall default value.
The form -D= will display a list of debugging options.
.SH COMMANDS
.PP
The following is the alphabetical list of recognized commands.
The jobcontrol or holdfile fields effective are listed as
part of the command.
.TP
?
.br
Help.
Also, any unrecognized input produces a help listing.
.TP
active [pr@[host]]
.br
makes a connection to the LPD server for the specified printer.
Closes the connection after making it.
This is useful for checking to see if a non-LPRng server is active.
.TP
abort { all | printer* }
.br
Kill the active job and disables unspooling.
The active job will not be deleted from the queue.
.br
Action: kills server process, updates spool control file
.I printing_disabled
field.
.TP
holdall { all | printer* }
.br
Turn on automatic job holding.
Jobs will be held until released with the hold command,
or holdall turned off with the noholdall command.
The :ah: printcap flag can enable automatic holding as well.
.br
Action: updates spool control file
.I holdall
field.
.TP
debug j all | printer } [string | off ]
.br
Set the debugging string for the specified printer.
This is a diagnostic aid and should be used with caution;
refer to the reference manual for details of the debugging string format.
In general,
you can use:
.br
.I INTEGER
\- the integer number sets the general debugging level
.br
keyword=value
.br
\- sets a debugging variable to a specified value
.br
keyword
\- sets a debugging flag
.br
Action: updates spool control file
.I debug
field.
.TP
disable { all | printer }
.br
Disable spooling to the specified spool queues.
See
.I enable.
.br
Action: updates spool control file
.I printing_disabled
field.
.TP
class {all | printer } (off | classlist | X=globmatch)
.br
Controls the class of jobs currently being printed.
The
.I off
option removes any class restrictions.
The classlist option is a list of classes;
for example A,B,C would release classes A, B, and C for printing.
The X=globmatch form performs a match against a control file entry
starting with the indicated capital letter.
For example,  J=*form1* would allow jobs which had the string
form1 in their job title to be printed.
.TP
down {all | printer }
.br
Disable both queuing and printing for all printers or the selected printer.
See
.I up.
.br
Action: updates spool control file
.I printing_disabled
and
.I spooling_disabled
fields.
.TP
enable { all | printer }
.br
Enable spooling for all printers or the selected printer.
.br
Action: updates spool control file
.I spooling_disabled
field.
.TP
exit or quit
.br
terminate LPC program
.TP
help
.br
Print a short help message.
.TP
hold printer { jobid }
.br
Holds the specified printer job (or jobs) in the queue.
The job will not be
printed even if spooling and printing is enabled. It must be released
with the relase command.
The holdall command
or :ah: printcap flag can enable automatic holding of spooled jobs.
.br
Action: updates job holdfile
.I hold
field with time that job was held.
A 0 value release job for printing.
.TP
kill { all | printer ... }
.br
Do an abort command followed by a start command.
This is a quick way to kill off a server that has problems.
Note that due to race conditions and delays,
that the start command may not be effective this action should
be monitored to ensure that the job restarts.
.TP
lpd [printer@[host]]
.br
determines if LPD daemon process on the print server is running,
and gets the PID.
This is handy to determine if the LPD daemon was killed or
aborted due to abnormal conditions.
.TP
lpq printer [options]
.br
Run
.I lpq
from inside the 
.I lpc
program.
.TP
lprm printer jobid [jobid]*
.br
Run
.I lprm
from inside the 
.I lpc
program.
Not all options are supported - this form requires
the printer name and jobid to be specified explicitly.
.TP
move printer { jobid }  destinationPrinter
.br
Send the specified jobs to the destination printer and remove them
from the printer queue.
.br
Action: updates the job holdfile
.I move
field with the destination and starts a server process to do the
job transfer.
.TP
noholdall { all | printer* }
.br
Turn off automatic job holding.
See holdall command.
.TP
quit or exit
.br
terminate LPC program
.TP
redirect [printer [destinationPrinter | off ]]
.br
redirect the jobs in the printer queue to another printer
or turn redirection off.
.br
Action: updates spool control file
.I redirect
field.
.TP
redo [printer [jobid]]
.br
Reprint the selected job.
.TP
release [printer [jobid]]
.br
Releases the selected job for printing.
.TP
reread [ printer [@host] ]
.br
Sends a request to the LPD server for the printer to reread the
configuration and printcap information.
This is equivalent to using
kill -HUP serverpid,
but can be done for servers on remote hosts.
.TP
start { all | printer }
.br
Start the printer.
This is useful when some abnormal condition causes the server to
terminate unexpectedly leaving jobs in the queue.
.I Lpq
will report that there is no daemon present when this condition occurs. 
.TP
status { all | printer }
Display the status of daemons and queues on the local machine.
.TP
stop { all | printer }
.br
Disable any further unspooling after the current job completes.
.TP
topq printer [ jobid ]
.br
Place the selected jobs at the top of the printer queue.
.TP
up {all | printer ...}
Enables queuing and printing for the specified or all printers.
Privileged.
.SH JOBIDS
.PP
The LPQ command displays a job identifier for each job,
which can be used in LPC commands to identify a specific job.
In the commands descriptions above,
the jobid can be a user name,
a job number,
a job identifier,
or a glob pattern.
The glob pattern is matched against the job identifier.
In command which have an optional jobid,
if none is specified then the first printable job in the queue is acted on.
.SH FILES
.PP
The files used by LPRng are set by values in the
printer configuration file.
The following are a commonly used set of default values.
.nf
.ta \w'/var/spool/lpd/printcap.<hostname>           'u
/etc/lpd.conf		LPRng configuration file
/etc/printcap		printer description file
/etc/lpd.perms	printer permissions
/var/spool/printer*		spool directories
/var/spool/printer*/printer	lock file for queue control
/var/spool/printer*/control.printer	queue control
/var/spool/printer*/active.printer	active job
/var/spool/printer*/log.printer	log file
.fi
.SH "SEE ALSO"
lpd.conf(5),
lpc(8),
lpd(8),
lpr(1),
lpq(1),
lprm(1),
printcap(5),
lpd.perms(5),
pr(1).
.SH DIAGNOSTICS
.nf
Most of the diagnostics are self explanatory.
If you are puzzled over the exact cause of failure,
set the debugging level on (-D5) and run again.
The debugging information will 
help you to pinpoint the exact cause of failure.
.fi
.SH "HISTORY"
.LP
LPRng is a enhanced printer spooler system
with functionality similar to the Berkeley LPR software.
In 1988 Patrick Powell released
the PLP (Public Line Printer) software,
which went through several evolutions.
Justin Mason (jmason@iona.ie)
generated PLP4.0 from several older releases of PLP.
In 1992 Patrick Powell
release LPRng,
a completely redesigned and newly written version of the software.
.LP
The LPRng mailing list is plp@iona.ie;
subscribe by sending mail to plp-request@iona.ie with
the word subscribe in the body.
The software is available from ftp://iona.ie/pub/LPRng.
.LP
LPRng is distributed under the GNU software license for non-commercial
use,
the Artistic License for limited commercial use. 
Commerical support and licensing is available through
Patrick Powell <papowell@sdsu.edu>.


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/lprng/lprng/src/common/lpc.c
/***************************************************************************
 * LPRng - An Extended Print Spooler System
 *
 * Copyright 1988-1999, Patrick Powell, San Diego, CA
 *     papowell@astart.com
 * See LICENSE for conditions of use.
 *
 ***************************************************************************/

 static char *const _id =
"$Id: lpc.c,v 1.8 2001-03-07 21:23:09 ghudson Exp $";


/***************************************************************************
 * SYNOPSIS
 *      lpc [ -PPrinter] [-S Server] [-V] [-D debug] [command]
 * commands:
 *
 *   status [printer]     - show printer status (default is all printers)
 *   msg   printer  ....  - printer status message
 *   start [printer]      - start printing
 *   stop [printer]       - stop printing
 *   up [printer]         - start printing and spooling
 *   down [printer]       - stop printing and spooling
 *   enable [printer]     - enable spooling
 *   disable [printer]    - disable spooling
 *   abort [printer]      - stop printing, kill server
 *   kill [printer]       - stop printing, kill server, restart printer
 *
 *   topq printer (user [@host] | host | jobnumer)*
 *   hold printer (all | user [@host] | host |  jobnumer)*
 *   release printer (all | user [@host] | host | jobnumer)*
 *
 *   lprm printer [ user [@host]  | host | jobnumber ] *
 *   lpq printer [ user [@host]  | host | jobnumber ] *
 *   lpd [pr | pr@host]   - PID of LPD server
 *   active [pr |pr@host] - check to see if server accepting connections
 *   local [all | pr ]     - show client configuration and printcap info 
 *   server [all |pr ]     - show server configuration and printcap info 
 *
 * DESCRIPTION
 *   lpc sends a  request to lpd(8)
 *   and reports the status of the command
 ****************************************************************************
 *
 * Implementation Notes
 * Patrick Powell Wed Jun 28 21:28:40 PDT 1995
 * 
 * The LPC program is an extremely simplified front end to the
 * LPC functionality in the server.  The commands send to the LPD
 * server have the following format:
 * 
 * \6printer user command options
 * 
 * If no printer is specified, the printer is the default from the
 * environment variable, etc.
 * 
 */

#include "lp.h"
#include "initialize.h"
#include "getprinter.h"
#include "sendreq.h"
#include "child.h"
#include "control.h"
#include "sendauth.h"
#include "getopt.h"
#include "patchlevel.h"
#include "errorcodes.h"

/**** ENDINCLUDE ****/


/***************************************************************************
 * main()
 * - top level of LPP Lite.
 *
 ****************************************************************************/


#undef EXTERN
#undef DEFINE
#define EXTERN
#define DEFINE(X) X
#include "lpc.h"

 void use_msg(void);
 void doaction( struct line_list *args );

int main(int argc, char *argv[], char *envp[])
{
	char *s;
	int i;
	char msg[ LINEBUFFER ];
	struct line_list args;


	/* set signal handlers */
	(void) plp_signal (SIGHUP, cleanup_HUP);
	(void) plp_signal (SIGINT, cleanup_INT);
	(void) plp_signal (SIGQUIT, cleanup_QUIT);
	(void) plp_signal (SIGTERM, cleanup_TERM);
	(void) plp_signal (SIGCHLD, SIG_DFL);

	/*
	 * set up the user state
	 */
#ifndef NODEBUG
	Debug = 0;
#endif

	Init_line_list(&args);
	Initialize(argc, argv, envp);
	Setup_configuration();

	/* scan the argument list for a 'Debug' value */

	Get_parms(argc, argv);      /* scan input args */

	DEBUG1("lpc: Optind '%d', argc '%d'", Optind, argc );
	if(DEBUGL1){
		int ii;
		for( ii = Optind; ii < argc; ++ii ){
			logDebug( " [%d] '%s'", ii, argv[ii] );
		}
	}


	if( Optind < argc ){
		for( i = Optind; argv[i]; ++i ){
			Add_line_list(&args,argv[i],0,0,0);
		}
		Check_max(&args,2);
		args.list[args.count] = 0;
		doaction( &args );
	} else while(1){
		fprintf( stdout, "lpc>" );
		fflush( stdout );
		if( fgets( msg, sizeof(msg), stdin ) == 0 ) break;
		if( (s = strchr( msg, '\n' )) ) *s = 0;
		DEBUG1("lpc: '%s'", msg );
		Free_line_list(&args);
		Split(&args,msg,Whitespace,0,0,0,0,0);
		Check_max(&args,2);
		args.list[args.count] = 0;
		if(DEBUGL1)Dump_line_list("lpc - args", &args );
		if( args.count == 0 ) continue;
		s = args.list[0];
		if( strcasecmp(s,"exit") == 0 || s[0] == 'q' || s[0] == 'Q' ){
			break;
		}
		doaction(&args);
	}
	Free_line_list(&args);
	Errorcode = 0;
	Is_server = 0;
	cleanup(0);
	return(0);
}

void doaction( struct line_list *args )
{
	int action, fd, n, i;
	struct line_list l;
	char msg[SMALLBUFFER];
	char *s, *t, *w, *printcap;

	Init_line_list(&l);
	s = t = w = printcap = 0;
	if( args->count == 0 ) return;
	action = Get_controlword( args->list[0] );
	if(DEBUGL1)Dump_line_list("doaction - args", args );
	if( action == 0 ){
		use_msg();
		return;
	}
	if( args->count > 1 ){
		Set_DYN(&Printer_DYN,args->list[1]);
		Fix_Rm_Rp_info();
		DEBUG1("doaction: Printer '%s', RemotePrinter '%s', RemoteHost '%s'",
			Printer_DYN, RemotePrinter_DYN, RemoteHost_DYN );
		if( (s = strchr(args->list[1],'@')) ) *s = 0;
		if( Server ){
			DEBUG1("doaction: overriding Remotehost with '%s'", Server );
			Set_DYN(&RemoteHost_DYN, Server );
		}
	} else if( Printer_DYN == 0 ){
		/* get the printer name */
		Get_printer();
		Fix_Rm_Rp_info();
		if( Server ){
			DEBUG1("doaction: overriding Remotehost with '%s'", Server );
			Set_DYN(&RemoteHost_DYN, Server );
		}
	}

	DEBUG1("lpc: RemotePrinter_DYN '%s', RemoteHost_DYN '%s'", RemotePrinter_DYN, RemoteHost_DYN );
	if( action == OP_SERVER ){
		Is_server = 1;
		Setup_configuration();
	}
	if( action == OP_LOCAL || action == OP_SERVER ){
		s = Join_line_list(&Config_line_list,"\n :");
		printcap = safestrdup3("Config","\n :",s,__FILE__,__LINE__);
		if( (w = strrchr(printcap,' ')) ) *w = 0;
		if( Write_fd_str( 1, printcap ) < 0 ) cleanup(0);
		if( s ) free(s); s = 0;
		if( t ) free(t); t = 0;
		if( printcap ) free(printcap); printcap = 0;

		if( args->count > 1 ){
			Get_all_printcap_entries();
			if( !strcasecmp(args->list[1], "all") ){
				if(s) free(s); s = 0;
				if( t ) free(t); t = 0;
				if(printcap) free(printcap); printcap = 0;
				s = Join_line_list(&PC_names_line_list,"\n :");
				printcap = safestrdup3("\nNames","\n :",s,__FILE__,__LINE__);
				if( (w = strrchr(printcap,' ')) ) *w = 0;
				if( Write_fd_str( 1, printcap ) < 0 ) cleanup(0);

				if(s) free(s); s = 0;
				if( t ) free(t); t = 0;
				if(printcap) free(printcap); printcap = 0;
				s = Join_line_list(&All_line_list,"\n :");
				printcap = safestrdup3("\nAll","\n :",s,__FILE__,__LINE__);
				if( (w = strrchr(printcap,' ')) ) *w = 0;
				if( Write_fd_str( 1, printcap ) < 0 ) cleanup(0);
			}

			if( Write_fd_str( 1,"\nPrintcap Information\n") < 0 ) cleanup(0);
			for( i = 0; i < All_line_list.count; ++i ){
				if( strcasecmp(args->list[1], "all")
					&& strcasecmp(args->list[1], All_line_list.list[i]) ){
					continue;
				}
				Set_DYN(&Printer_DYN,All_line_list.list[i]);
				Fix_Rm_Rp_info();
				if( s ) free(s); s = 0;
				if( t ) free(t); t = 0;
				if( printcap ) free(printcap); printcap = 0;
				t = Join_line_list(&PC_alias_line_list,"|");
				s = Join_line_list(&PC_entry_line_list,"\n :");
				if( s && t ){
					if( (w = strrchr(t,'|')) ) *w = 0;
					printcap = safestrdup3(t,"\n :",s,__FILE__,__LINE__);
					if( (w = strrchr(printcap,' ')) ) *w = 0;
					if( Write_fd_str( 1, printcap ) < 0 ) cleanup(0);
				}
			}
		}
		if( s ) free(s); s = 0;
		if( t ) free(t); t = 0;
		if( printcap ) free(printcap); printcap = 0;
	} else if( action == OP_LPQ || action == OP_LPRM ){
		pid_t pid, result;
		plp_status_t status;
		if( args->count == 1 && Printer_DYN ){
			plp_snprintf(msg,sizeof(msg), "-P%s", Printer_DYN );
			Add_line_list(args,msg,0,0,0);
			Check_max(args,1);
			args->list[args->count] = 0;
		} else if( args->count > 1 ){
			s = args->list[1];
			if( strcasecmp(s,"all") ){
				plp_snprintf(msg,sizeof(msg), "-P%s", s );
			} else {
				strcpy(msg, "-a" );
			}
			if( s ) free(s);
			args->list[1] = safestrdup(msg,__FILE__,__LINE__);
		}
		if(DEBUGL1)Dump_line_list("ARGS",args);
		if( (pid = dofork(0)) == 0 ){
			/* we are going to close a security loophole */
			Full_user_perms();
			/* this would now be the same as executing LPQ as user */
			close_on_exec(3);
			execvp( args->list[0],args->list );
			Diemsg( _("execvp failed - '%s'"), Errormsg(errno) );
			exit(0);
		} else if( pid < 0 ) {
			Diemsg( _("fork failed - '%s'"), Errormsg(errno) );
		}
		while(1){
			result = plp_waitpid( pid, &status, 0 );
			if( (result == -1 && errno != EINTR) || result == 0 ) break;
		}
		DEBUG1("lpc: system pid %d, exit status %s",
			result, Decode_status( &status ) );
	} else {
		Add_line_list(&l, Logname_DYN, Value_sep, 0, 0 );
		Add_line_list(&l, args->list[0], Value_sep, 0, 0);
		Remove_line_list(args, 0);
		if( args->count > 0 ) {
			Add_line_list(&l, RemotePrinter_DYN, Value_sep, 0, 0 );
			Remove_line_list(args, 0);
		}
		Merge_line_list(&l, args, 0, 0, 0 );
		Check_max(&l, 1 );
		l.list[l.count] = 0;
		fd = Send_request( 'C', REQ_CONTROL, l.list, Connect_timeout_DYN,
			Send_query_rw_timeout_DYN, 1 );
		if( fd > 0 ){
			while( (n = read(fd, msg, sizeof(msg))) > 0 ){
				if( (write(1,msg,n)) < 0 ) cleanup(0);
			}
		}
		close(fd);
	}
	Free_line_list(&l);
}

/* VARARGS2 */
#ifdef HAVE_STDARGS
 void setstatus (struct job *job,char *fmt,...)
#else
 void setstatus (va_alist) va_dcl
#endif
{
#ifndef HAVE_STDARGS
    struct job *job;
    char *fmt;
#endif
	char msg[LARGEBUFFER];
    VA_LOCAL_DECL

    VA_START (fmt);
    VA_SHIFT (job, struct job * );
    VA_SHIFT (fmt, char *);

	msg[0] = 0;
	if( Verbose ){
		(void) plp_vsnprintf( msg, sizeof(msg)-2, fmt, ap);
		strcat( msg,"\n" );
		if( Write_fd_str( 2, msg ) < 0 ) cleanup(0);
	}
	VA_END;
	return;
}

 void send_to_logger (struct job *job,const char *header, char *fmt){;}
/* VARARGS2 */
#ifdef HAVE_STDARGS
 void setmessage (struct job *job,const char *header, char *fmt,...)
#else
 void setmessage (va_alist) va_dcl
#endif
{
#ifndef HAVE_STDARGS
    struct job *job;
    char *fmt, *header;
#endif
	char msg[LARGEBUFFER];
    VA_LOCAL_DECL

    VA_START (fmt);
    VA_SHIFT (job, struct job * );
    VA_SHIFT (header, char * );
    VA_SHIFT (fmt, char *);

	msg[0] = 0;
	if( Verbose ){
		(void) plp_vsnprintf( msg, sizeof(msg)-2, fmt, ap);
		strcat( msg,"\n" );
		if( Write_fd_str( 2, msg ) < 0 ) cleanup(0);
	}
	VA_END;
	return;
}


/***************************************************************************
 * void Get_parms(int argc, char *argv[])
 * 1. Scan the argument list and get the flags
 * 2. Check for duplicate information
 ***************************************************************************/

 void usage(void);

 char LPC_optstr[] 	/* LPC options */
 = "D:P:S:V";

/* scan the input arguments, setting up values */

void Get_parms(int argc, char *argv[] )
{
	int option;

	while ((option = Getopt (argc, argv, LPC_optstr )) != EOF) {
		switch (option) {
		case 'D': /* debug has already been done */
			Parse_debug( Optarg, 1 );
			break;
		case 'P': if( Optarg == 0 ) usage();
			Set_DYN(&Printer_DYN,Optarg); break;
		case 'V':
			Verbose = !Verbose;
			break;
		case 'S':
			Server = Optarg;
			break;
		default:
			usage();
		}
	}
	if( Verbose ) {
		fprintf( stderr, _("Version %s, Copyright 1988-1999 Patrick Powell\n"), PATCHLEVEL );
	}
}

 char *msg = N_("\
usage: %s [-A] [-Ddebuglevel] [-Pprinter] [-V] [command]\n\
 with no commands, reads from stdin\n\
  -Pprinter    - specify printer\n\
  -V           - increase information verbosity\n\
  -Ddebuglevel - debug level\n\
 commands:\n\
 active    (printer[@host])        - check for active server\n\
 abort     (printer[@host] | all)  - stop server\n\
 class     printer[@host] (class | off)      - show/set class printing\n\
 disable   (printer[@host] | all)  - disable queueing\n\
 debug     (printer[@host] | all) debugparms - set debug level for printer\n\
 down      (printer[@host] | all)  - disable printing and queueing\n\
 enable    (printer[@host] | all)  - enable queueing\n\
 hold      (printer[@host] | all) (name[@host] | job | all)*   - hold job\n\
 holdall   (printer[@host] | all)  - hold all jobs on\n\
 kill      (printer[@host] | all)  - stop and restart server\n\
 lpd       (printer[@host]) - get LPD PID \n\
 lpq       (printer[@host] | all) (name[@host] | job | all)*   - invoke LPQ\n\
 lprm      (printer[@host] | all) (name[@host]|host|job| all)* - invoke LPRM\n\
 msg printer message text  - set status message\n\
 move printer (user|jobid)* target - move jobs to new queue\n\
 noholdall (printer[@host] | all)  - hold all jobs off\n\
 printcap  (printer[@host] | all)  - report printcap values\n\
 quit                              - exit LPC\n\
 redirect  (printer[@host] | all) (printer@host | off )*       - redirect jobs\n\
 redo      (printer[@host] | all) (name[@host] | job | all)*   - release job\n\
 release   (printer[@host] | all) (name[@host] | job | all)*   - release job\n\
 reread    (printer[@host])        - LPD reread database information\n\
 start     (printer[@host] | all)  - start printing\n\
 status    (printer[@host] | all)  - status of printers\n\
 stop      (printer[@host] | all)  - stop  printing\n\
 topq      (printer[@host] | all) (name[@host] | job | all)*   - reorder job\n\
 up        (printer[@host] | all) - enable printing and queueing\n\
   diagnostic:\n\
      defaultq               - show default queue for LPD server\n\
      local  (printer | all) - client config and printcap information\n\
      server (printer | all) - server config and printcap\n");

void use_msg(void)
{
	fprintf( stderr, _(msg), Name );
}
void usage(void)
{
	use_msg();
	exit(1);
}

int Start_worker( struct line_list *args, int fd )
{
	if(DEBUGL1)Dump_line_list("LPC - Dummy Start_worker",args);
	Errorcode = JABORT;
	fatal(LOG_ERR,"LPC - Dummy Start_worker called");
	return(fd);
}

#if TEST

#include "permission.h"
#include "gethostinfo.h"
#include "lpd.h"

int Send_request(
	int class,					/* 'Q'= LPQ, 'C'= LPC, M = lprm */
	int format,					/* X for option */
	char **options,				/* options to send */
	int connect_timeout,		/* timeout on connection */
	int transfer_timeout,		/* timeout on transfer */
	int output					/* output on this FD */
	)
{
	int i, n;
	int socket = 1;
	char cmd[SMALLBUFFER];

	cmd[0] = format;
	cmd[1] = 0;
	plp_snprintf(cmd+1, sizeof(cmd)-1, "%s", RemotePrinter_DYN);
	for( i = 0; options[i]; ++i ){
		n = strlen(cmd);
		plp_snprintf(cmd+n,sizeof(cmd)-n," %s",options[i] );
	}
	Perm_check.remoteuser = Logname_DYN;
	Perm_check.user = Logname_DYN;
	Perm_check.remotehost = &Localhost_IP;
	Is_server = 1;
	Job_control(&socket,cmd);
	return(-1);
}

#endif
